src/cephadm/cephadmlib/node_proxy/main.py

```diff
@@ -0,0 +1,93 @@
+from threading import Thread
+from .redfishdellsystem import RedfishDellSystem
+from .reporter import Reporter
+from .util import Config, Logger
+from typing import Dict, Any, Optional
+import traceback
+
+DEFAULT_CONFIG = {
+    'reporter': {
+        'check_interval': 5,
+        'push_data_max_retries': 30,
+        'endpoint': 'https://127.0.0.1:7150/node-proxy/data',
+    },
+    'system': {
+        'refresh_interval': 5
+    },
+    'server': {
+        'port': 8080,
+    },
+    'logging': {
+        'level': 20,
+    }
+}
+
+
+class NodeProxyFetchOobError(Exception):
+    pass
+
+
+class NodeProxy(Thread):
+    def __init__(self, **kw: Dict[str, Any]) -> None:
+        super().__init__()
+        for k, v in kw.items():
+            setattr(self, k, v)
+        self.exc: Optional[Exception] = None
+        # self.cp_shutdown_event = Event()
+        self.log = Logger(__name__)
+
+    def run(self) -> None:
+        try:
+            self.main()
+        except Exception as e:
+            self.exc = e
+            return
+
+    def shutdown(self) -> None:
+        self.log.logger.info('Shutting down node-proxy...')
+        self.system.client.logout()
+        self.system.stop_update_loop()
+        self.reporter_agent.stop()
+
+    def check_auth(self, realm: str, username: str, password: str) -> bool:
+        return self.__dict__['username'] == username and \
+            self.__dict__['password'] == password
+
+    def check_status(self) -> bool:
+        if self.__dict__.get('system') and not self.system.run:
+            raise RuntimeError('node-proxy encountered an error.')
+        if self.exc:
+            traceback.print_tb(self.exc.__traceback__)
+            self.log.logger.error(f'{self.exc.__class__.__name__}: {self.exc}')
+            raise self.exc
+        return True
+
+    def main(self) -> None:
+        # TODO: add a check and fail if host/username/password/data aren't passed
+        self.config = Config('/etc/ceph/node-proxy.yml', default_config=DEFAULT_CONFIG)
+        self.log = Logger(__name__, level=self.config.__dict__['logging']['level'])
+        self.cephx = self.__dict__['cephx']
```

Comment:
is it necessary to access the kwargs in this way, given `__init__` already did
```
        for k, v in kw.items():
            setattr(self, k, v)
```
Note: I'm assuming all of these `foo = self.__dict__['bar']` are for accessing kwargs, so correct me if they're not.

Reply:
L69 seems useless, indeed.
The other calls `self.__dict__['foo']` were to make mypy/flake8 stop complain.

Reply:
I ended up explicitly setting the different attributes so the code is more readable and clearer.